src/device/sound.rs

```diff
@@ -0,0 +1,1563 @@
+        let mut jack_features_str = String::new();
+        bitflags::parser::to_writer(&jack_features, &mut jack_features_str).unwrap();
+        let connected_status = if self.connected == 1 {
+            "CONNECTED"
+        } else {
+            "DISCONNECTED"
+        };
+        write!(
+            f,
+            "features: {}, hda_reg_defconf: {}, hda_reg_caps: {}, connected: {}",
+            jack_features_str, self.hda_reg_defconf, self.hda_reg_caps, connected_status
+        )
+    }
+}
+
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes)]
+struct VirtIOSndJackInfoRsp {
+    hdr: VirtIOSndHdr,
+    body: VirtIOSndJackInfo,
+}
+
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes)]
+struct VirtIOSndJackRemap {
+    hdr: VirtIOSndJackHdr,
+    association: u32,
+    sequence: u32,
+}
+
+#[repr(C)]
+#[derive(AsBytes, FromBytes, FromZeroes)]
+struct VirtIOSndPcmHdr {
+    /// specifies request type (VIRTIO_SND_R_PCM_*)
+    hdr: VirtIOSndHdr,
+    /// specifies a PCM stream identifier from 0 to streams - 1
+    stream_id: u32,
+}
+
+enum PcmStreamFeatures {
+    VirtioSndPcmFShmemHost = 0,
+    VirtioSndPcmFShmemGuest,
+    VirtioSndPcmFMsgPolling,
+    VirtioSndPcmFEvtShmemPeriods,
+    VirtioSndPcmFEvtXruns,
+}
+
+impl From<PcmStreamFeatures> for u32 {
+    fn from(value: PcmStreamFeatures) -> Self {
+        value as _
+    }
+}
+
+enum PcmSampleFormat {
+    /* analog formats (width / physical width) */
+    VirtioSndPcmFmtImaAdpcm = 0, /* 4 / 4 bits */
+    VirtioSndPcmFmtMuLaw,        /* 8 / 8 bits */
+    VirtioSndPcmFmtALaw,         /* 8 / 8 bits */
+    VirtioSndPcmFmtS8,           /* 8 / 8 bits */
+    VirtioSndPcmFmtU8,           /* 8 / 8 bits */
+    VirtioSndPcmFmtS16,          /* 16 / 16 bits */
+    VirtioSndPcmFmtU16,          /* 16 / 16 bits */
+    VirtioSndPcmFmtS18_3,        /* 18 / 24 bits */
+    VirtioSndPcmFmtU18_3,        /* 18 / 24 bits */
+    VirtioSndPcmFmtS20_3,        /* 20 / 24 bits */
+    VirtioSndPcmFmtU20_3,        /* 20 / 24 bits */
+    VirtioSndPcmFmtS24_3,        /* 24 / 24 bits */
+    VirtioSndPcmFmtU24_3,        /* 24 / 24 bits */
+    VirtioSndPcmFmtS20,          /* 20 / 32 bits */
+    VirtioSndPcmFmtU20,          /* 20 / 32 bits */
+    VirtioSndPcmFmtS24,          /* 24 / 32 bits */
+    VirtioSndPcmFmtU24,          /* 24 / 32 bits */
+    VirtioSndPcmFmtS32,          /* 32 / 32 bits */
+    VirtioSndPcmFmtU32,          /* 32 / 32 bits */
+    VirtioSndPcmFmtFloat,        /* 32 / 32 bits */
+    VirtioSndPcmFmtFloat64,      /* 64 / 64 bits */
+    /* digital formats (width / physical width) */
+    VirtioSndPcmFmtDsdU8,          /* 8 / 8 bits */
+    VirtioSndPcmFmtDsdU16,         /* 16 / 16 bits */
+    VirtioSndPcmFmtDsdU32,         /* 32 / 32 bits */
+    VirtioSndPcmFmtIec958Subframe, /* 32 / 32 bits */
+}
+
+impl From<PcmSampleFormat> for u64 {
+    fn from(value: PcmSampleFormat) -> Self {
+        value as _
+    }
+}
+
+enum PcmFrameRate {
```

Comment:
This doesn't seem to be used anywhere, remove it.

Reply:
Removed.